shared/src/main/scala/fetch.scala

```diff
@@ -148,52 +151,53 @@ object `package` {
     def map2[A, B, C](f: (A, B) => C)(fa: Fetch[A], fb: Fetch[B]): Fetch[C] =
       Fetch.join(fa, fb).map({ case (a, b) => f(a, b) })
 
+    private[this] def simplify(results: DataSourceCache): (FetchOp ~> FetchOp) = {
+      new (FetchOp ~> FetchOp) {
+        def apply[B](f: FetchOp[B]): FetchOp[B] = f match {
+          case one @ FetchOne(id, ds) => {
+              results
+                .get(ds.identity(id))
+                .fold(one: FetchOp[B])(b => Cached(b).asInstanceOf[FetchOp[B]])
+            }
+          case many @ FetchMany(ids, ds) => {
+              val fetched = ids.map(id => results.get(ds.identity(id))).unwrap.sequence
+              fetched.fold(many: FetchOp[B])(results => Cached(results))
+            }
+          case conc @ Concurrent(manies) => {
+              val newManies = manies
+                .filterNot({ fm =>
+                  val ids: NonEmptyList[Any] = fm.as.asInstanceOf[NonEmptyList[Any]]
+                  val ds: DataSource[Any, _] = fm.ds.asInstanceOf[DataSource[Any, _]]
+
+                  ids
+                    .map(id => {
```

Comment:
Nothing wrong with this but the I'd and block it's verbose. May be simplified like the next chained call